### PR TITLE
feat: Assignment 16 Design Pattern Singleton and Factory Pattern

### PR DESCRIPTION
**Pull request by Subhasri-V-P**
**Background: **
To understand singleton and factory pattern.

**Implementation: **
1. LoggingSystem Class following Singleton pattern is created.
2. Using Factory pattern an abstract class with methods are created and the method is implemented in concrete class.
3.  An interface ILogger is created with log method.
4. PlainTextLogger and JSONLogger classes implements ILogger interface.
5. The singleton pattern and factory pattern are integrated by LogMessage() method in LoggingSystem class.
6. LogMessageManager asks user choice and performs operation requested by the user.

### CODE REVIEW

File: src/LoggingSystem/LogMessageManager.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    Console.WriteLine($"{(int)operations}. {operations}");
+                }
+
+                Console.WriteLine("Choose type of logging to be done: ");
+                int userChoice = Enum.GetNames(typeof(OperationsToPerform)).Length.GetValidUserChoice();
+                switch (userChoice)
+                {
+                    case 1:
+                        LoggingSystem loggingSystem = LoggingSystem.Instance;
```
Comment (Shaalini-Subrammanian) [inline src/LoggingSystem/LogMessageManager.cs:42]: Can we have it outside the switch case

File: src/LoggingSystem/PlainTextLogger.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    /// </summary>
+    public class PlainTextLogger : ILogger
+    {
+        /// <summary>
+        /// Log method - logs message as plain text
+        /// </summary>
+        /// <param name="message">Message to be logged</param>
+        public void Log(string message)
+        {
+            StreamWriter streamWriterObject = new (Path.Combine(Directory.GetCurrentDirectory().ToString(), "plaintextFile.txt"), true);
```
Comment (Shaalini-Subrammanian) [inline src/LoggingSystem/PlainTextLogger.cs:18]:
```suggestion
            StreamWriter streamWriterObject = new (Path.Combine(Directory.GetCurrentDirectory().ToString(), "LoggerTextFile.txt"), true);
```

File: src/LoggingSystem/JSONLogger.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    /// </summary>
+    public class JSONLogger : ILogger
+    {
+        /// <summary>
+        /// Log method - logs as json in the file
+        /// </summary>
+        /// <param name="message">Message to be logged</param>
+        public void Log(string message)
+        {
+            StreamWriter streamWriterObject = new (Path.Combine(Directory.GetCurrentDirectory().ToString(), "jsonFile.json"), true);
```
Comment (Shaalini-Subrammanian) [inline src/LoggingSystem/JSONLogger.cs:18]:
```suggestion
            StreamWriter streamWriterObject = new (Path.Combine(Directory.GetCurrentDirectory().ToString(), "LoggerJsonFile.json"), true);
```

**Review by Shaalini-Subrammanian — APPROVED** [review]

File: tests/unit/LoggingSystemTests/LoggingSystemTests/LoggingSystemTests.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            // Arrange
+            LoggingSystem loggingSystem = LoggingSystem.Instance;
+            string message = "Hello";
+            string type = "PlainTextLogger";
+
+            // Act
+            loggingSystem.LogMessage(
+                message,
+                type);
+            var mockFileObject = new StreamReader(Path.Combine(Directory.GetCurrentDirectory().ToString(), "plaintextFile.txt"));
```
Comment (Shaalini-Subrammanian) [inline tests/unit/LoggingSystemTests/LoggingSystemTests/LoggingSystemTests.cs:19]: UpdateFile name

File: src/LoggingSystem/LogMessageManager.cs (hunk 1 above)
Comment (Subhasri-V-P) [reply src/LoggingSystem/LogMessageManager.cs:42]: Instance is declared outside switch case. The previous declaration was to check the hash code.

File: src/LoggingSystem/PlainTextLogger.cs (hunk 2 above)
Comment (Subhasri-V-P) [reply src/LoggingSystem/PlainTextLogger.cs:18]: File name is refactored.

File: src/LoggingSystem/JSONLogger.cs (hunk 3 above)
Comment (Subhasri-V-P) [reply src/LoggingSystem/JSONLogger.cs:18]: File name refactored.

File: tests/unit/LoggingSystemTests/LoggingSystemTests/LoggingSystemTests.cs (hunk 4 above)
Comment (Subhasri-V-P) [reply tests/unit/LoggingSystemTests/LoggingSystemTests/LoggingSystemTests.cs:19]: File names updated in tests.

**Pull request merged**